src/WinList/WinList.man
.\" t
.\" @(#)WinList.1	002/07/00
.TH WinList 1.8 "Feb 7 2000"
.UC
.SH NAME
\fBWinList\fP \- the AfterStep window list module

.SH SYNOPSIS
\fBWinList\fP [\-f \fIconfig_file\fP] [\-v | \--version] [\-h | \--help]
.sp
Although \fBWinList\fP can be run from the command line, it is normally 
spawned by afterstep in the \fIautoexec\fP file, in the following manner:
.nf
.sp
	Module	"I"	WinList
.sp
.fi
From within an \fIafterstep\fP menu file by 'WinList':
.nf
.sp
% cat "~/GNUstep/Library/AfterStep/start/Modules/WinList"
	Module "WinList" WinList
	MiniPixmap "mini-as.xpm"
.fi

To stop the \fBWinList\fP module, a menu item can be used:
.nf
.sp
% cat "~/GNUstep/Library/AfterStep/start/Modules/WinList_Stop"
	KillModuleByName "Stop WinList" WinList
	MiniPixmap "mini-as.xpm"
.fi

.SH DESCRIPTION
The \fBWinList\fP module provides a window list made up of buttons, each
corresponding to a window that AfterStep is managing.  Clicking on the
buttons with any of the three mouse buttons will either do a default
action or can be user configured by editing the \fIwinlist\fP file.

.SH OPTIONS
.IP "\-f \fIconfig_file\fP"
Use \fIconfig_file\fP instead of the default configuration file.

.IP "\-h | \--help"
Print a usage message and exit successfully.

.IP "\-v | \--version"
Print version information and exit successfully.

.SH INITIALIZATION
During initialization, \fBWinList\fP will search a configuration file
which describes the colors, font, actions to take on mouse clicks, and
geometry settings to use. This file will be the one specified by the 
\fI\-f\fP option (if given).  Else, the ~/GNUstep/Library/AfterStep/winlist 
file will be used if it exists.  Else, /usr/local/share/afterstep/winlist 
will be used.

If the \fBWinList\fP executable is linked to another name, ie 'ln -s
\fBWinList\fP \fBOtherWinList\fP', then another module called
\fBOtherWinList\fP can be started, with a completely different set of
configurations than \fBWinList\fP, simply by adding \fI*OtherWinList\fP
options following the same \fI*WinList\fP configuration options below to the
\fIwinlist\fP file. This way multiple Identify programs can be used.

.SH INVOCATION
\fBWinList\fP can be invoked from the \fIautoexec\fP file or from a menu
item as shown in the Synopsis section above.

\fBWinList\fP can also be invoked from binding "Module WinList
\fIargument\fP" to a keystroke or mouse button using \fITransient\fP as an
argument which will cause \fBWinList\fP to resemble the built-in window
list.

AfterStep will search through the directories specified in the
\fIModulePath\fP configuration in the \fIbase.#bpp\fP (where # is your
colordepth) to attempt to locate WinList.

.SH CONFIGURATION OPTIONS
\fBWinList\fP reads the \fIwinlist\fP file when it starts up, and looks
for its configuration options.  The MyStyle structure should now be specified
in the look files.  This allows for a greater control of the running module
(different WinList look with different desktop looks, dynamic reconfiguration
etc.).

.IP "MyStyle ""*WinList"""
.IP "  \fIstyle_option\fP"
.IP "~MyStyle"

The possible \fIstyle_option\fPs follow:

.RS
.IP "Font \fIfont\fP"
Specifies the font associated with this style. \fIfont\fP is a standard 
X11 font definition.

.IP "ForeColor \fIcolor\fP"
Specifies the text color associated with this style. \fIcolor\fP is a 
standard X11 color definition.

.IP "BackColor \fIcolor\fP"
Specifies the background color associated with this style. \fIcolor\fP 
is a standard X11 color definition.

.IP "TextStyle \fIstyle\fP"
Specifies the text \fIstyle\fP associated with this style. \fIstyle\fP 
can be 0, 1, or 2:

.nf
0: normal text
1: 3d effect #1
2: 3d effect #2
.fi

.IP "MaxColors \fImaximum\fP"
Specifies the \fImaximum\fP number of colors that the \fIBackGradient\fP 
can use. jpegs specified with \fIBackPixmap\fP will also be limited to 
this number of colors.

.IP "BackGradient \fItype\fP \fIfrom\fP \fIto\fP"
Specifies that a gradient should be used as a background instead of a 
solid color. The gradient colors start at \fIfrom\fP and end at \fIto\fP. 
\fIfrom\fP and \fIto\fP are standard X11 color definitions. \fItype\fP 
can be 1, 2, 3, 4, or 5:

.nf
1: Wharf-style diagonal gradient
2: Horizontal from top to bottom
3: Horizontal from top/bottom to center
4: Vertical from left to right
5: Vertical from left/right to center
.fi

.IP "BackPixmap \fItype\fP \fIpixmap_name|color_name\fP"
Specifies that a pixmap should be used as a background instead of a 
solid color. This option has two forms, depending on the value of 
\fItype\fP. If \fItype\fP is 128, the second parameter, \fIpixmap_name\fP 
must be the name of an xpm, jpeg, or png image which can be found in 
\fIPixmapPath\fP. If \fItype\fP is 129, the root pixmap (desktop 
background) will be used, and the optional second parameter must be a 
color with which the root pixmap will be shaded.

.nf
128: tiled pixmap from the upper left
129: "transparent"
.fi

.RE

.IP "*WinListGeometry \fIgeometry\fP"
Specifies the location and gravity of the WinList window.  At the current
time, size is not supported and WinList will resize itself as buttons are
added.  If the NoAnchor option is not specified then the window's gravity
corner will be anchored, and the window will grow in the opposite direction.
(i.e. If the geometry is specified -5-5, that is SoutEastGravity.  This will
cause the window to draw up and to the left as windows are added)

.IP "*WinListHideGeometry \fIgeometry\fP"
Specifies the location of the WinList window when there are no windows. 
It will cause WinList to display nice little window with AfterStep version
in there. If size is supplied WinList will center the label within the 
window.

.IP "*WinListUseSkipList"
Tells \fBWinList\fP to not show the windows that are have
\fIWindowListSkip\fP as a \fIStyle\fP option in the \fdatabase\fP file.

.IP "*WinListNoAnchor"
By default, \fBWinList\fP will anchor the gravity corner so the window will
grow in the opposite direction.  This undoes that option.

.IP "*WinListUseIconNames"
Tells WinList to use the icon name of the window instead of the full window
name.  This is useful to keep the width of the window small.

.IP "*WinListAction \fIaction response[,reponse...]\fP"
Tells WinList to do \fIresponse\fP when \fIaction\fP is done.  The
currently supported \fIaction\fPs are: Click1, Click2, Click3.  The
currently supported \fIresponse\fPs are any AfterStep built-in commands,
including modules and functions.

.IP "*WinListOrientation \fIdown|across\fP"
Will make a vertical or horizontal winlist.

.IP "*WinListMaxWidth \fIwidth in pixel\fP"
Sets either maxwidth for a vertical winlist or the actual width for a
horizontal "taskbar".

.IP "*WinListJustify \fI<Left|Center|Right>\fP"
If used, will justify the text as specifyed. If unspecified, the default is
to center text.

.IP "*WinListAutoHide \fInumber\fP"
Makes \fBWinList\fP auto-hide, when the mouse cursor travels off \fBWinList\fP it reduces its size and becomes a line, when the mouse cursor travels onto hidden \fBWinList\fP it readjusts its size to normal dimensions.  0 turns this feature off, positive numbers turn it on.  This option makes sense only when \fBWinList\fP is set to \fIStaysOnTop\fP in the database file. It is possible to turn \fIAutoHide\fP on/off when the module is running: pressing 'u' while the cursor is in the \fBWinList\fP window permanently unhides, pressing 'h' while the cursor is in the \fBWinList\fP window turns \fIAutoHide\fP on.

.IP "*WinListBalloons"
Turns on the balloon feature. A "balloon" is displayed showing the name 
of the button the pointer is resting on in the \fBWinList\fP. The balloon 
background and text are configured with the MyStyle "*WinListBalloon".  
See \fBafterstep(1x)\fP for more information on MyStyles.

.IP "*WinListBalloonBorderWidth \fInumber\fP"
Sets the width of the balloon window's border. Defaults to 1.

.IP "*WinListBalloonBorderColor \fIcolor\fP"
Sets the color of the balloon window's border. Defaults to black.

.IP "*WinListBalloonDelay \fInumber\fP"
Specifies the period in milliseconds to delay the appearance of the balloon.

.IP "*WinListBalloonYOffset \fInumber\fP"
The balloon window is positioned to be horizontally centered against
the \fBWinList\fP window it is describing. The vertical position may be
set as an offset. Negative offsets of \fI-N\fP are placed \fIN\fP
pixels above the \fBWinList\fP window, positive offsets of \fI+N\fP are 
placed \fIN\fP pixels below. Defaults to +2.

.SH FILES

.IP \fI/usr/local/share/afterstep/winlist\fP
The global configuration file for system wide defaults of all users.

.IP \fI$HOME/GNUstep/Library/AfterStep/winlist\fP
This file should be copied from the \fI/usr/local/share/afterstep/winlist\fP and edited to suit the user's specific requirements.  This file, if exists, will override the system wide default file.

.SH EXAMPLES
The \fIwinlist\fP file shipped contains a working example, refer to the
\fBFILES\fP section above for its configuration options.

.sp
To pop up the window list in tranient mode (appear only when button 3 is
pressed & held):
.sp
	...the \fIfeel\fP file would contain \-
.nf
	Mouse 3  R  A  Module "WinList" WinList Transient
.sp
.fi

.SH AUTHORS
Rene Fichter <ceezaer@cyberspace.org>
.nf
Mike Finger <mfinger@mermaid.micro.umn.edu>
.nf
David Mihm <dmihm@swbell.net> [Man page]
.nf
Rafal Wierzbicki <rafal@mcss.mcmaster.ca>
.SH SEE ALSO
.BR afterstep (1),


ChangeLog

2003-09-25  sasha

	* generating ChangeLog

2003-09-24  sasha

	* Completed look.Glass ; renamed PointerFore, PointerBack to
	  CursorFore, CursorBack to match old Cursor option; Moved Cursor
	  definitions from feel into look Completed Colorscheme's support
	  for Cursor's color

2003-09-23  sasha

	* Added look options for mouse pointer coloring: PointerFore and
	  PointerBack; Repaired code to update window's decorations when
	  stickiness flag is toggled;

2003-09-22  sasha

	* Added cursors to default feel file; enable cursor switching while
	  hovering over different elements of the interface; Added code to
	  change color of the cursor according to what colorscheme is set
	  to - may need to tweak a bit , or make colors configurable
	* Fixed locale parsing from LANG variable; Fixed module startup to
	  pass on current locae specified on command line

2003-09-19  sasha

	* added debugging code for cursor loading

2003-09-16  sasha

	* fixed bug with windows teleporting back to original position, if
	  moved/resized too fast

2003-09-15  sasha

	* updated configure scripts
	* Fixed configure.in to properly enable i18n in libAfterImage as
	  requested. Internationalized text should be displayed properly
	  now, even if its not in UTF8; Update src/test to work with new
	  libs and added test code for i18n testing
	* Fixed unquoted string in i18N code reported by Alex Mezin

2003-09-14  sasha

	* Added checks for locale and proper initialization of
	  libAfterImage t3ext drawing in accordance to its value; Fixed
	  problem with EdgeScroll being mangled when just look changes;
	  Fixed recently introduced bug causing wierd corners in framned
	  windows upon the look change
	* Fixed typoe for msg type in SendString being declared as some
	  wierd type of unsigned long long  - possibly causing wierdness on
	  64 bit machines

2003-09-13  sasha

	* implemented feedback functionality via sending the e-mail to
	  sasha :)

2003-09-12  sasha

	* added configure option to send e-mail back to developer on
	  installation; Fixed canvas border width handling while resizing
	  and in Pager;

2003-09-11  sasha

	* fixed handling of WM_PROTOCOL changes where WM_COMMAND would get
	  lost and session management would stop working; Found bug with
	  huge impact - all the geometry calculations were off as we were
	  forgetting border width of the frame window; Fixed handling of
	  button presses on border of the frame

2003-09-10  sasha

	* fixed NeXTish colorscheme somewhat
	* Fixed workspace state save/restore to allow terms to exec command
	  using -e, and to have argv's with spaces in them; Fixed frame
	  corner sizing to make it same width as is tbar size;

2003-09-09  sasha

	* More fixing Pager's geometry; Added -i cmd line parameter to
	  ascompose to allow for 'include' files (such as colorschemes,
	  etc.); Updated configure scripts to properly detect headers when
	  its run from libAfterBase; Updated version to beta2; Updated TEAM
	  file

2003-09-08  sasha

	* Fixed Pager to properly account for label size while sizing
	  min-desktop windows

2003-09-07  sasha

	* somewhat debugged Pager's geometry handling - need to add some
	  code to properly recalculate Pager's geometry if its soecified as
	  ratio of the main desktop vs. exact geometry

2003-09-03  sasha

	* added new AfterStep log icon; fixed wharf config file

2003-09-02  sasha

	* added new xml tag : <color> to define symbolic color names
	* changed handling of tiled images to clip it properly when
	  Right/Left align is used

2003-08-30  sasha

	* debugged workspace state handling code, and improved Wait loop

2003-08-29  sasha

	* mostly implemented new workspace state saving and restoring

2003-08-28  sasha

	* fixed workspace status generator to not save xrm params, as noone
	  seems to support thoise
	* added caching of pixmaps for unfocused bas in window decorations
	  to speed up fast focus switching
	* 64-bit fixes: CARD32 instead of Window type
	* 64-bit fixes: properties and strlen warnings
	* 64-bit fixes: converted intermodule communications to always use
	  32bit values
	* more 64 bit fixes

2003-08-27  sasha

	* 64-bit compilation fixes
	* Fixed slowness in intermodule communications manifesting itself
	  when desktops are switched
	* fixed handling of non-default MyBackgrounds, and proper adding of
	  DeskConfigs while switching desks
	* synching
	*  fixed segfaulting in Pager on look switches or after window
	  restacking
	* Fixed handling of hints flags - it was doing nothing before if
	  flags == 0 - was screwing MOtif Hints like totatlly; Fixed shaped
	  window handling in WinList; Added check to not do too much
	  redrawing when WinLIst is receiving window list - may want to do
	  the same in Pager

2003-08-26  sasha

	*  added parameters to import funcions, to allow for sized
	  image/preview stuff; Removed compression buffer from ASImage
	  structure to save up some memory; Hopefully fixed hints handling
	  on 64 bit hw

2003-08-24  sasha

	* fixed event handling when windows is getting mapped - no more
	  lost events due to not selected mask ; Fixed initial status when
	  mouse pointer is out of the screen; Fixed ICCCM size hints
	  handling to have BaseSize and MinSize interchange each other when
	  one is ommited - fixes GIMPs initial placement

2003-08-23  sasha

	* added new 3 icons to complete the set of default icons: switched
	  AfterStep to use ~/.afterstep instead of ridiculous ~/GNUstel/L/A
	  stuff

2003-08-22  sasha

	* Fixed handling of shaped windows - shape is not erroneously
	  cleared, and client's shape is subtracted from decoration's shape
	* fixed autodetection of zipped xpms and added 6 more icons to
	  default iconset

2003-08-21  sasha

	* more improvements to asimbrowser

2003-08-20  sasha

	* fixed makefile for ascp, when fltk is not found; added better
	  Style for BitchX to database; remnoved scrollbar from BitchX in
	  Wharf

src/Config/WinList.c
/*
 * Copyright (c) 2002 Sasha Vasko <sasha at aftercode.net>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.   See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */


#include "../../configure.h"

#include "../../include/asapp.h"
#include "../../include/afterstep.h"
#include "../../include/mystyle.h"
#include "../../include/parser.h"
#include "../../include/confdefs.h"
#include "../../include/balloon.h"

/*****************************************************************************
 *
 * This routine is responsible for reading and parsing the config file
 *
 ****************************************************************************/

extern SyntaxDef AlignSyntax;

TermDef       WinListTerms[] = {
    {0, "Geometry", 8, TT_GEOMETRY, WINLIST_Geometry_ID, NULL},
    {0, "MinSize", 7, TT_GEOMETRY, WINLIST_MinSize_ID, NULL},
    {0, "MaxSize", 7, TT_GEOMETRY, WINLIST_MaxSize_ID, NULL},
    {0, "MaxRows", 7, TT_INTEGER, WINLIST_MaxRows_ID, NULL},
    {0, "MaxColumns", 10, TT_INTEGER, WINLIST_MaxColumns_ID, NULL},
    {0, "MaxColWidth", 11, TT_INTEGER, WINLIST_MaxColWidth_ID, NULL},
    {0, "MinColWidth", 11, TT_INTEGER, WINLIST_MinColWidth_ID, NULL},
    {0, "UseName", 7, TT_INTEGER, WINLIST_UseName_ID, NULL},
    {0, "Align", 5, TT_FLAG, WINLIST_Align_ID, &AlignSyntax},
    {0, "Bevel", 5, TT_FLAG, WINLIST_Bevel_ID, &BevelSyntax},
    {TF_DONT_SPLIT, "Action", 6, TT_TEXT, WINLIST_Action_ID, NULL},
    {0, "UnfocusedStyle", 14, TT_TEXT, WINLIST_UnfocusedStyle_ID, NULL},
    {0, "FocusedStyle", 12, TT_TEXT, WINLIST_FocusedStyle_ID, NULL},
    {0, "StickyStyle", 11, TT_TEXT, WINLIST_StickyStyle_ID, NULL},
    {0, "FillRowsFirst", 13, TT_FLAG, WINLIST_FillRowsFirst_ID, NULL},
    {0, "UseSkipList", 11, TT_FLAG, WINLIST_UseSkipList_ID, NULL},
    {0, "ShapeToContents", 15, TT_FLAG, WINLIST_ShapeToContents_ID, NULL},
/* including MyStyles definitions processing */
	INCLUDE_MYSTYLE,

/* now special cases that should be processed by it's own handlers */
    BALLOON_TERMS,
    {0, NULL, 0, 0, 0}
};

SyntaxDef     WinListSyntax = {
	'\n',
	'\0',
	WinListTerms,
	0,										   /* use default hash size */
    ' ',
	"",
	"\t",
	"WinList configuration",
	NULL,
	0
};


WinListConfig *
CreateWinListConfig ()
{
	WinListConfig *config = (WinListConfig *) safecalloc (1, sizeof (WinListConfig));

    init_asgeometry (&(config->geometry));
    config->gravity = NorthWestGravity;
	config->max_rows = 1;
	config->show_name_type = ASN_Name;
    config->name_aligment = ALIGN_CENTER;
    config->balloon_conf = NULL;

	return config;
}

void
DestroyWinListConfig (WinListConfig * config)
{
	int           i;

	if (config->unfocused_style)
		free (config->unfocused_style);
	if (config->focused_style)
		free (config->focused_style);
	if (config->sticky_style)
		free (config->sticky_style);

	for (i = 0; i < MAX_MOUSE_BUTTONS; ++i)
		if (config->mouse_actions[i])
			free (config->mouse_actions[i]);

    Destroy_balloonConfig (config->balloon_conf);
	DestroyFreeStorage (&(config->more_stuff));
	DestroyMyStyleDefinitions (&(config->style_defs));

	free (config);
}

void
PrintWinListConfig (WinListConfig * config)
{
	int           i;

	fprintf (stderr, "WinListConfig = %p;\n", config);
	if (config == NULL)
		return;

	fprintf (stderr, "WinListConfig.RowsFirst = %s;\n",
			 get_flags (config->flags, WINLIST_FillRowsFirst) ? "True" : "False");
	fprintf (stderr, "WinListConfig.UseSkipList = %s;\n",
			 get_flags (config->flags, WINLIST_UseSkipList) ? "True" : "False");
	fprintf (stderr, "WinListConfig.set_flags = 0x%lX;\n", config->set_flags);
    fprintf (stderr, "WinListConfig.geometry.flags = 0x%X;\n", config->geometry.flags);
    fprintf (stderr, "WinListConfig.geometry = %+d%+d;\n", config->geometry.x, config->geometry.y);
	fprintf (stderr, "WinListConfig.gravity = %d;\n", config->gravity);
	fprintf (stderr, "WinListConfig.MinSize = %dx%d;\n", config->min_width, config->min_height);
	fprintf (stderr, "WinListConfig.MaxSize = %dx%d;\n", config->max_width, config->max_height);
	fprintf (stderr, "WinListConfig.MaxRows = %d;\n", config->max_rows);
	fprintf (stderr, "WinListConfig.MaxColumns = %d;\n", config->max_columns);
	fprintf (stderr, "WinListConfig.min_col_width = %d;\n", config->min_col_width);
	fprintf (stderr, "WinListConfig.max_col_width = %d;\n", config->max_col_width);

	fprintf (stderr, "WinListConfig.show_name_type = %d;\n", config->show_name_type);	/* 0, 1, 2, 3 */
    fprintf (stderr, "WinListConfig.name_aligment = %lX;\n", config->name_aligment);
    fprintf (stderr, "WinListConfig.bevel = %lX;\n", config->bevel);

	fprintf (stderr, "WinListConfig.unfocused_style = %p;\n", config->unfocused_style);
	if (config->unfocused_style)
		fprintf (stderr, "WinListConfig.unfocused_style = \"%s\";\n", config->unfocused_style);
	fprintf (stderr, "WinListConfig.focused_style = %p;\n", config->focused_style);
	if (config->focused_style)
		fprintf (stderr, "WinListConfig.focused_style = \"%s\";\n", config->focused_style);
	fprintf (stderr, "WinListConfig.sticky_style = %p;\n", config->sticky_style);
	if (config->sticky_style)
		fprintf (stderr, "WinListConfig.sticky_style = \"%s\";\n", config->sticky_style);

	for (i = 0; i < MAX_MOUSE_BUTTONS; ++i)
	{
		fprintf (stderr, "WinListConfig.mouse_action[%d] = %p;\n", i, config->mouse_actions[i]);
		if (config->mouse_actions[i])
			fprintf (stderr, "WinListConfig.mouse_action[%d] = \"%s\";\n", i, config->mouse_actions[i]);
	}
}

WinListConfig *
ParseWinListOptions (const char *filename, char *myname)
{
	ConfigDef    *ConfigReader =
        InitConfigReader (myname, &WinListSyntax, CDT_Filename, (void *)filename, NULL);
	WinListConfig *config = CreateWinListConfig ();

	FreeStorageElem *Storage = NULL, *pCurr;
	ConfigItem    item;
	MyStyleDefinition **styles_tail = &(config->style_defs);

	if (!ConfigReader)
		return config;

	item.memory = NULL;
	PrintConfigReader (ConfigReader);
	ParseConfig (ConfigReader, &Storage);
	PrintFreeStorage (Storage);

	/* getting rid of all the crap first */
	StorageCleanUp (&Storage, &(config->more_stuff), CF_DISABLED_OPTION);

    config->balloon_conf = Process_balloonOptions (Storage, NULL);

    for (pCurr = Storage; pCurr; pCurr = pCurr->next)
	{
		if (pCurr->term == NULL)
			continue;

		if (pCurr->term->type == TT_FLAG)
        {
            switch (pCurr->term->id)
            {
                case WINLIST_FillRowsFirst_ID:
                    SET_CONFIG_FLAG (config->flags, config->set_flags, WINLIST_FillRowsFirst);
                    break;
                case WINLIST_UseSkipList_ID:
                    SET_CONFIG_FLAG (config->flags, config->set_flags, WINLIST_UseSkipList);
                    break;
                case WINLIST_UseIconNames_ID:
                    if (!get_flags (config->set_flags, WINLIST_UseName))
                    {
                        set_flags (config->set_flags, WINLIST_UseName);
                        config->show_name_type = ASN_IconName;
                    }
                    break;
                case WINLIST_Align_ID :
                    set_flags( config->set_flags, WINLIST_Align );
                    config->name_aligment = ParseAlignOptions( pCurr->sub );
                    break ;
                case WINLIST_Bevel_ID :
                    set_flags( config->set_flags, WINLIST_Bevel );
                    config->bevel = ParseBevelOptions( pCurr->sub );
                    break ;
                case WINLIST_ShapeToContents_ID:
                    SET_CONFIG_FLAG (config->flags, config->set_flags, WINLIST_ShapeToContents);
                    break;
            }
        }else
		{
			if (!ReadConfigItem (&item, pCurr))
				continue;

			switch (pCurr->term->id)
			{
                case WINLIST_Geometry_ID:
                    set_flags (config->set_flags, WINLIST_Geometry);
                    config->geometry = item.data.geometry ;
                    break;
                case WINLIST_MinSize_ID:
                    set_flags (config->set_flags, WINLIST_MinSize);
                    if (get_flags (item.data.geometry.flags, WidthValue))
                        config->min_width = item.data.geometry.width;
                    else
                        config->min_width = 0;
                    if (get_flags (item.data.geometry.flags, HeightValue))
                        config->min_height = item.data.geometry.height;
                    else
                        config->min_height = 0;
                    break;
                case WINLIST_MaxWidth_ID:
                    set_flags (config->set_flags, WINLIST_MaxSize);
                    config->max_width = item.data.integer;
                    break;
                case WINLIST_MaxSize_ID:
                    set_flags (config->set_flags, WINLIST_MaxSize);
                    if (get_flags (item.data.geometry.flags, WidthValue))
                        config->max_width = item.data.geometry.width;
                    else
                        config->max_width = 0;
                    if (get_flags (item.data.geometry.flags, HeightValue))
                        config->max_height = item.data.geometry.height;
                    else
                        config->max_height = 0;
                    break;
                case WINLIST_MaxRows_ID:
                    set_flags (config->set_flags, WINLIST_MaxRows);
                    config->max_rows = item.data.integer;
                    break;
                case WINLIST_MaxColumns_ID:
                    set_flags (config->set_flags, WINLIST_MaxColumns);
                    config->max_columns = item.data.integer;
                    break;
                case WINLIST_MinColWidth_ID:
                    set_flags (config->set_flags, WINLIST_MinColWidth);
                    config->min_col_width = item.data.integer;
                    break;
                case WINLIST_MaxColWidth_ID:
                    set_flags (config->set_flags, WINLIST_MaxColWidth);
                    config->max_col_width = item.data.integer;
                    break;
                case WINLIST_UseName_ID:
                    set_flags (config->set_flags, WINLIST_UseName);
                    config->show_name_type = item.data.integer % ASN_NameTypes;
                    break;
                case WINLIST_Action_ID:
                    {
                        char         *ptr = item.data.string;
                        int           action_no = 0, i;

                        if (mystrncasecmp (ptr, "Click", 5) == 0)
                            ptr += 5;
                        if (isdigit (ptr[0]))
                        {
                            action_no = atoi (ptr);
                            if (action_no <= 0)
                                action_no = 1;
                            --action_no;
                            action_no %= MAX_MOUSE_BUTTONS;
                            i = 0;
                            while (!isspace (ptr[i]) && ptr[i])
                                ++i;
                            while (isspace (ptr[i]) && ptr[i])
                                ++i;
                            ptr += i;
                        }
                        if (*ptr)
                        {
                            ptr = mystrdup (ptr);
                            REPLACE_STRING (config->mouse_actions[action_no], ptr);
                        }
                        item.ok_to_free = 1;
                    }
                    break;
                case WINLIST_UnfocusedStyle_ID:
                    REPLACE_STRING (config->unfocused_style, item.data.string);
                    break;
                case WINLIST_FocusedStyle_ID:
                    REPLACE_STRING (config->focused_style, item.data.string);
                    break;
                case WINLIST_StickyStyle_ID:
                    REPLACE_STRING (config->sticky_style, item.data.string);
                    break;
                case MYSTYLE_START_ID:
                    styles_tail = ProcessMyStyleOptions (pCurr->sub, styles_tail);
                    item.ok_to_free = 1;
                    break;
                case WINLIST_Orientation_ID:
                    if (mystrncasecmp (item.data.string, "across", 6) == 0)
                        SET_CONFIG_FLAG (config->flags, config->set_flags, WINLIST_FillRowsFirst);
                    item.ok_to_free = 1;
                    break;
                default:
                    item.ok_to_free = 1;
			}
		}
	}
	ReadConfigItem (&item, NULL);
	DestroyConfig (ConfigReader);
	DestroyFreeStorage (&Storage);
/*    PrintMyStyleDefinitions( config->style_defs ); */
	return config;

}


#if 0
/* returns:
 *            0 on success
 *              1 if data is empty
 *              2 if ConfigWriter cannot be initialized
 *
 */
int
WriteWinListOptions (const char *filename, char *myname, WinListConfig * config, unsigned long flags)
{
	ConfigDef    *PagerConfigWriter = NULL;
	FreeStorageElem *Storage = NULL, **tail = &Storage;
	char         *Decorations = NULL;

	if (config == NULL)
		return 1;
	if ((PagerConfigWriter = InitConfigWriter (myname, &PagerSyntax, CDT_Filename, (void *)filename)) == NULL)
		return 2;

	CopyFreeStorage (&Storage, config->more_stuff);

    if (config->balloon_conf)
		tail = balloon2FreeStorage (&WinListSyntax, tail, config->balloon_conf);

	/* building free storage here */
	/* geometry */
	tail = Geometry2FreeStorage (&PagerSyntax, tail, &(config->geometry), PAGER_GEOMETRY_ID);
	/* icon_geometry */
	tail = Geometry2FreeStorage (&PagerSyntax, tail, &(config->icon_geometry), PAGER_ICON_GEOMETRY_ID);
	/* labels */
	tail = StringArray2FreeStorage (&PagerSyntax, tail, config->labels, desk1, desk2, PAGER_LABEL_ID);
	/* styles */
#ifdef PAGER_BACKGROUND
	tail = StringArray2FreeStorage (&PagerSyntax, tail, config->styles, desk1, desk2, PAGER_STYLE_ID);
#endif
	/* align */
	tail = Integer2FreeStorage (&PagerSyntax, tail, config->align, PAGER_ALIGN_ID);
	/* flags */
	if (!(config->flags & REDRAW_BG))
		tail = Flag2FreeStorage (&PagerSyntax, tail, PAGER_DRAW_BG_ID);
	if (config->flags & START_ICONIC)
		tail = Flag2FreeStorage (&PagerSyntax, tail, PAGER_START_ICONIC_ID);
	if (config->flags & FAST_STARTUP)
		tail = Flag2FreeStorage (&PagerSyntax, tail, PAGER_FAST_STARTUP_ID);
	if (config->flags & SET_ROOT_ON_STARTUP)
		tail = Flag2FreeStorage (&PagerSyntax, tail, PAGER_SET_ROOT_ID);
	if (config->flags & STICKY_ICONS)
		tail = Flag2FreeStorage (&PagerSyntax, tail, PAGER_STICKY_ICONS_ID);

	/* small_font_name */
	tail = String2FreeStorage (&PagerSyntax, tail, config->small_font_name, PAGER_SMALL_FONT_ID);
	/* rows */
	tail = Integer2FreeStorage (&PagerSyntax, tail, config->rows, PAGER_ROWS_ID);
	/* columns */
	tail = Integer2FreeStorage (&PagerSyntax, tail, config->columns, PAGER_COLUMNS_ID);

	/* now storing PagerDecorations */
	{
		ConfigDef    *DecorConfig = InitConfigWriter (myname, &PagerDecorationSyntax, CDT_Data, NULL);
		FreeStorageElem *DecorStorage = NULL, **d_tail = &DecorStorage;

		/* flags */
		if (!(config->flags & ~USE_LABEL))
			d_tail = Flag2FreeStorage (&PagerDecorationSyntax, d_tail, PAGER_DECOR_NOLABEL_ID);
		if (!(config->flags & PAGE_SEPARATOR))
			d_tail = Flag2FreeStorage (&PagerDecorationSyntax, d_tail, PAGER_DECOR_NOSEPARATOR_ID);
		if (!(config->flags & SHOW_SELECTION))
			d_tail = Flag2FreeStorage (&PagerDecorationSyntax, d_tail, PAGER_DECOR_NOSELECTION_ID);
		if (config->flags & LABEL_BELOW_DESK)
			d_tail = Flag2FreeStorage (&PagerDecorationSyntax, d_tail, PAGER_DECOR_LABEL_BELOW_ID);
		if (config->flags & HIDE_INACTIVE_LABEL)
			d_tail = Flag2FreeStorage (&PagerDecorationSyntax, d_tail, PAGER_DECOR_HIDE_INACTIVE_ID);

		/* selection_color */
		d_tail = String2FreeStorage (&PagerDecorationSyntax, d_tail, config->selection_color, PAGER_DECOR_SEL_COLOR_ID);
		/* border_width */
		d_tail =
			Integer2FreeStorage (&PagerDecorationSyntax, d_tail, config->border_width, PAGER_DECOR_BORDER_WIDTH_ID);
		/* grid_color */
		d_tail = String2FreeStorage (&PagerDecorationSyntax, d_tail, config->grid_color, PAGER_DECOR_GRID_COLOR_ID);
		/* border_color */
		d_tail = String2FreeStorage (&PagerDecorationSyntax, d_tail, config->border_color, PAGER_DECOR_BORDER_COLOR_ID);

		WriteConfig (DecorConfig, &DecorStorage, CDT_Data, (void **)&Decorations, 0);
		DestroyConfig (DecorConfig);
		if (DecorStorage)
			DestroyFreeStorage (&DecorStorage);

		tail = String2FreeStorage (&PagerSyntax, tail, Decorations, PAGER_DECORATION_ID);
		if (Decorations)
		{
			free (Decorations);
			Decorations = NULL;
		}
	}

	/* writing config into the file */
	WriteConfig (PagerConfigWriter, &Storage, CDT_Filename, (void **)&filename, flags);
	DestroyConfig (PagerConfigWriter);

	if (Storage)
	{
		fprintf (stderr, "\n%s:Config Writing warning: Not all Free Storage discarded! Trying again...", myname);
		DestroyFreeStorage (&Storage);
		fprintf (stderr, (Storage != NULL) ? " failed." : " success.");
	}
	return 0;
}

#endif
